### PR TITLE
eliminate `now` error at a fixed rate schedule executor

### PR DESCRIPTION
**Pull request by dmsolr**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

to eliminate `System.currentTimeMillis()` error at a fixed rate schedule executor. That will cause `CounterWindow` cannot to remove the element as expected.

### CODE REVIEW

**Comment by hanahmily** [comment]
Would you debug some issues? Please elaborate what's these changes' purpose. It looks like this should not be merged until finding the root cause.

**Comment by wu-sheng** [comment]
@dmsolr Could you share an error about `System.currentTimeMillis()`? If this could go wrong, then the agent has a higher risk than backend. Also, we never received a report about this

**Comment by dmsolr** [comment]
Sorry for maybe I use the improper word about `error`.
The period of calling `System.currentTimeMillis()` twice may not equal to `period of fetcher`.

for example, the period of fetching = 10s, we can get the time list as follows,
1624265391580
1624265401585
1624265411581
1624265421582

I set the duration as `PT10S`, and then CounterWindow needs to remove the `1624265401585` after `1624265411581`.
But it is not removed because of `1624265411581 - 1624265401585 = 9996 < 10000`.

**Comment by wu-sheng** [comment]
> I set the duration as PT10S, and then CounterWindow needs to remove the 1624265401585 after 1624265411581.
But it is not removed because of 1624265411581 - 1624265401585 = 9996 < 10000.

Wouldn't remove that in the next iteration?

**Comment by dmsolr** [comment]
It is sure will be removed in the next iteration. But it will lead to aggregation computing improperly and an element(1624265411581) does not remove in this iteration.

https://github.com/apache/skywalking/blob/09944e00a95a61b2a091087ccb49f8b6a66f0604/oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java#L47-L58

**Comment by wu-sheng** [comment]
>  But it will lead to aggregation computing improperly and an element(1624265411581) does not remove in this iteration.

I can't see this as a removal issue. If you define a window of 5 slots(last 5 minutes), even there are 6 slots kept in short term, there is no point `rate` function that has an issue. If there is an issue, it is a rate function error, rather than a removal bug.

**Comment by wu-sheng** [comment]
@hanahmily An update, this issue is caused by currently, `increase` and `rate` are peeking the first and last of the windows to calculate the value. If the oldest data can't be removed in time, then the result is wrong. 

I have asked @dmsolr to fix in the way, that always use the duration to calculate the number of slot for rate/increase, and don't always peek data for the last element, but `latest` - `N slot`(from rate/increase function parameter)

**Review by hanahmily — APPROVED** [review]
> @hanahmily An update, this issue is caused by currently, `increase` and `rate` are peeking the first and last of the windows to calculate the value. If the oldest data can't be removed in time, then the result is wrong.
> 
> I have asked @dmsolr to fix in the way, that always use the duration to calculate the number of slot for rate/increase, and don't always peek data for the last element, but `latest` - `N slot`(from rate/increase function parameter)

Got > @hanahmily An update, this issue is caused by currently, `increase` and `rate` are peeking the first and last of the windows to calculate the value. If the oldest data can't be removed in time, then the result is wrong.
> 
> I have asked @dmsolr to fix in the way, that always use the duration to calculate the number of slot for rate/increase, and don't always peek data for the last element, but `latest` - `N slot`(from rate/increase function parameter)

Thanks for your elaboration, got the point. These changes look great to me.

**Review by wu-sheng — CHANGES_REQUESTED** [review]
@hanahmily What I asked, has not been done. This PR is manipulate time, which is not recommended. Let's wait for slot base mathematics.

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java [hunk 1]
```diff
@@ -47,15 +47,20 @@
...
+            windows.put(id, new PriorityQueue<>());
         }
+
         Queue<Tuple2<Long, Double>> window = windows.get(id);
         window.offer(Tuple.of(now, value));
-        Tuple2<Long, Double> ps = window.element();
-        if ((now - ps._1) >= windowSize) {
-            window.remove();
+
+        long waterLevel = now - windowSize + 60;
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java:56]: What does `60` mean?

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java [hunk 2]
```diff
@@ -47,15 +47,27 @@
...
+        long waterLevel = now - windowSize;
+        Tuple2<Long, Double> peek = window.peek();
+        if (peek._1 > waterLevel) {
+            return peek;
+        }
+        Tuple2<Long, Double> result = peek;
+        while (peek._1 <= waterLevel) {
+            result = window.poll();
+            peek = window.element();
         }
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java:65]: This logic seems correct, have you tried `Collections.binarySearch()`? `window` is an ordered list, right? If so, a binary search would be much faster than a sequential search。

File: oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java [hunk 3]
```diff
@@ -42,20 +42,46 @@
...
+            return peek;
+        }
+
+        Tuple2<Long, Double> result = peek;
+        while (peek._1 < waterLevel) {
+            result = window.poll();
+            peek = window.element();
+        }
+
+        if (waterLevel - result._1 <= peek._1 - waterLevel) {
```
Comment (wu-sheng) [inline oap-server/analyzer/meter-analyzer/src/main/java/org/apache/skywalking/oap/meter/analyzer/dsl/counter/CounterWindow.java:68]:
```suggestion
        // Choose the closed slot to the expected timestamp
        if (waterLevel - result._1 <= peek._1 - waterLevel) {
```

**Review by wu-sheng — APPROVED** [review]
LGTM

**Review by hanahmily — APPROVED** [review]
LGTM good fix!

**Pull request merged**